static/app/views/organizationIntegrations/sentryAppExternalForm.tsx

```diff
@@ -0,0 +1,323 @@
+export type Config = {
+  uri: string;
+  required_fields?: FieldFromSchema[];
+  optional_fields?: FieldFromSchema[];
+};
+
+// only need required_fields and optional_fields
+type State = Omit<Config, 'uri'> & {
+  optionsByField: Map<string, Array<{label: string; value: any}>>;
+};
+
+type Props = {
+  api: Client;
+  sentryAppInstallation: SentryAppInstallation;
+  appName: string;
+  config: Config;
+  action: 'create' | 'link';
+  element: 'issue-link' | 'alert-rule-action';
+  extraFields?: {[key: string]: any};
+  extraRequestBody?: {[key: string]: any};
+  onSubmitSuccess: (...params: any[]) => void;
+  getFieldDefault?: (field: FieldFromSchema) => string;
+};
+
+export class SentryAppExternalForm extends Component<Props, State> {
+  state: State = {optionsByField: new Map()};
+
+  componentDidMount() {
+    this.resetStateFromProps();
+  }
+
+  componentDidUpdate(prevProps: Props) {
+    if (prevProps.action !== this.props.action) {
+      this.model.reset();
+      this.resetStateFromProps();
+    }
+  }
+
+  model = new FormModel();
+
+  // reset the state when we mount or the action changes
+  resetStateFromProps() {
+    const {config, action, extraFields} = this.props;
+    this.setState({
+      required_fields: config.required_fields,
+      optional_fields: config.optional_fields,
+    });
+    // we need to pass these fields in the API so just set them as values so we don't need hidden form fields
+    this.model.setInitialData({
+      ...extraFields,
+      action,
+      uri: config.uri,
+    });
+  }
+
+  onSubmitError = () => {
+    const {action, appName} = this.props;
+    addErrorMessage(t('Unable to %s %s %s.', action, appName, this.getElementText()));
+  };
+
+  getOptions = (field: FieldFromSchema, input: string) =>
+    new Promise(resolve => {
+      this.debouncedOptionLoad(field, input, resolve);
+    });
+
+  getElementText = () => {
+    const {element} = this.props;
+    switch (element) {
+      case 'issue-link':
+        return 'issue';
+      case 'alert-rule-action':
+        return 'alert';
+      default:
+        return 'connection';
+    }
+  };
+
+  debouncedOptionLoad = debounce(
+    // debounce is used to prevent making a request for every input change and
+    // instead makes the requests every 200ms
+    async (field: FieldFromSchema, input, resolve) => {
+      const choices = await this.makeExternalRequest(field, input);
+      const options = choices.map(([value, label]) => ({value, label}));
+      const optionsByField = new Map(this.state.optionsByField);
+      optionsByField.set(field.name, options);
+      this.setState({
+        optionsByField,
+      });
+      return resolve(options);
+    },
+    200,
+    {trailing: true}
+  );
+
+  makeExternalRequest = async (field: FieldFromSchema, input: FieldValue) => {
+    const install = this.props.sentryAppInstallation;
+    const {extraRequestBody = {}} = this.props;
+    const query: {[key: string]: any} = {
+      ...extraRequestBody,
```

Comment:
As stated above, the Issue Linking API requires projectId to be passed. This allow the parent component specify this.